hpx/parallel/spmd_block.hpp

```diff
@@ -0,0 +1,156 @@
+//  Copyright (c) 2017 Antoine Tran Tan
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_PARALLEL_SPMD_BLOCK_HPP)
+#define HPX_PARALLEL_SPMD_BLOCK_HPP
+
+#include <hpx/include/parallel_for_each.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/lcos/barrier.hpp>
+/*#include <hpx/runtime/actions/lambda_to_action.hpp>*/
+#include <hpx/runtime/launch_policy.hpp>
+#include <hpx/runtime/naming/name.hpp>
+/*#include <hpx/runtime/serialization/serialize.hpp>*/
+
+#include <boost/range/irange.hpp>
+
+#include <cstddef>
+#include <memory>
+#include <string>
+#include <type_traits>
+
+namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v2)
+{
+    namespace detail
+    {
+        template <typename T>
+        struct extract_first_parameter
+        {};
+
+        // Specialization for lambdas
+        template <typename ClassType, typename ReturnType>
+        struct extract_first_parameter<ReturnType(ClassType::*)() const>
+        {
+            using type = std::false_type;
+        };
+
+        // Specialization for lambdas
+        template <typename ClassType,
+            typename ReturnType, typename Arg0, typename... Args>
+        struct extract_first_parameter<
+            ReturnType(ClassType::*)(Arg0, Args...) const>
+        {
+            using type = Arg0;
+        };
+    }
+
+    /// The class spmd_block defines an interface for launching
+    /// multiple images while giving handles to each image to interact with
+    /// the remaining images. The \a define_spmd_block function templates create
+    /// multiple images of a user-defined lambda and launches them in a possibly
+    /// separate thread. A temporary spmd block object is created and diffused
+    /// to each image. The constraint for the lambda given to the
+    /// define_spmd_block function is to accept a spmd_block as first parameter.
+    struct spmd_block
+    {
+        spmd_block(){}
+
+        spmd_block(
+            std::string name, std::size_t num_images, std::size_t image_id)
+        : name_(name), num_images_(num_images), image_id_(image_id)
```

Comment:
`name` has been provisionnaly removed (for local tests purposes) in d2ea96b42cd8313d36c082c9797b43a4c516abe4